hudi-common/src/main/java/org/apache/hudi/common/table/TableSchemaResolver.java

```diff
@@ -295,91 +295,19 @@ private MessageType convertAvroSchemaToParquet(Schema schema) {
+   * @param prevSchema previous instance of the schema
+   * @param newSchema new instance of the schema
    */
-  public static boolean isSchemaCompatible(Schema oldSchema, Schema newSchema) {
-    if (oldSchema.getType() == newSchema.getType() && newSchema.getType() == Schema.Type.RECORD) {
-      // record names must match:
-      if (!SchemaCompatibility.schemaNameEquals(newSchema, oldSchema)) {
-        return false;
-      }
-
-      // Check that each field in the oldSchema can populated the newSchema
-      for (final Field oldSchemaField : oldSchema.getFields()) {
-        final Field newSchemaField = SchemaCompatibility.lookupWriterField(newSchema, oldSchemaField);
-        if (newSchemaField == null) {
-          // C4 or C2: newSchema does not correspond to any field in the oldSchema
-          return false;
-        } else {
-          if (!isSchemaCompatible(oldSchemaField.schema(), newSchemaField.schema())) {
-            // C5: The fields do not have a compatible type
-            return false;
-          }
-        }
-      }
-
-      // Check that new fields added in newSchema have default values as they will not be
-      // present in oldSchema and hence cannot be populated on reading records from existing data.
-      for (final Field newSchemaField : newSchema.getFields()) {
-        final Field oldSchemaField = SchemaCompatibility.lookupWriterField(oldSchema, newSchemaField);
-        if (oldSchemaField == null) {
-          if (newSchemaField.defaultVal() == null) {
-            // C3: newly added field in newSchema does not have a default value
-            return false;
-          }
-        }
-      }
-
-      // All fields in the newSchema record can be populated from the oldSchema record
-      return true;
-    } else {
-      // Use the checks implemented by Avro
-      // newSchema is the schema which will be used to read the records written earlier using oldSchema. Hence, in the
-      // check below, use newSchema as the reader schema and oldSchema as the writer schema.
-      org.apache.avro.SchemaCompatibility.SchemaPairCompatibility compatResult =
-          org.apache.avro.SchemaCompatibility.checkReaderWriterCompatibility(newSchema, oldSchema);
-      return compatResult.getType() == org.apache.avro.SchemaCompatibility.SchemaCompatibilityType.COMPATIBLE;
-    }
+  public static boolean isSchemaCompatible(Schema prevSchema, Schema newSchema) {
+    // NOTE: We're establishing compatibility of the {@code prevSchema} and {@code newSchema}
+    //       as following: {@code newSchema} is considered compatible to {@code prevSchema},
+    //       iff data written using {@code prevSchema} could be read by {@code newSchema}
+    SchemaCompatibility.SchemaPairCompatibility result =
+        SchemaCompatibility.checkReaderWriterCompatibility(newSchema, prevSchema);
+    return result.getType() == SchemaCompatibility.SchemaCompatibilityType.COMPATIBLE;
```

Comment:
Basically the schema check is relaxed here without field-by-field comparison, with only read compatibility check.

Reply:
Discussed offline: previously we had custom implementation that was specifically disallowing columns being dropped. Since we're relaxing this constraint in this PR there's no reason for us to continue maintain that custom logic. Instead we can solely rely on Avro schema compatibility checking.